[Style] Convert the '-webkit-line-box-contain' property to strong style types

weinig opened the pull request:
#### 9fe5de14695d6cf5e5bd4eab7e0cc85af7e58ae0
<pre>
[Style] Convert the &apos;-webkit-line-box-contain&apos; property to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=302203">https://bugs.webkit.org/show_bug.cgi?id=302203</a>

Reviewed by Darin Adler.

Converts the &apos;-webkit-line-box-contain&apos; property to use strong style types.

* Source/WTF/wtf/EnumSet.h:
    - Exposes the underlying enum type via a `using value_type` declaration.
      Needed for use in generic contexts.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new file.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
    - Add mappings macro for `Style::WebkitLineBoxContainValue`.

* Source/WebCore/css/CSSProperties.json:
    - Update style converter used by &apos;-webkit-line-box-contain&apos; to the
      new style type.

* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
    - Adds new aggregates `SpaceSeparatedEnumSet` and `CommaSeparatedEnumSet`
      that wrap `WTF::EnumSet` with specified serialization and conversion.
    - Adds new wrapper, `EnumSetOrKeywordBase` which works like `ValueOrKeyword`
      and `ListOrNone`, except specialized for an `EnumSet` value, meaning it
      has exposed iterators, contains functions, and toRaw/fromRaw functions.

* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBoxBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
* Source/WebCore/rendering/LegacyRootInlineBox.cpp:
* Source/WebCore/rendering/updating/RenderTreeBuilderFirstLetter.cpp:
    - Update for new type name for `-webkit-line-box-contain` values.

* Source/WebCore/rendering/style/RenderStyleDifference.h:
    - Replaces `LOG_IF_DIFFERENT_WITH_CAST_AND_CONSTRUCTION` and `LOG_RAW_OPTIONSET_IF_DIFFERENT`
      with a single `LOG_IF_DIFFERENT_WITH_FROM_RAW`. This means that types using
      `LOG_RAW_OPTIONSET_IF_DIFFERENT` now have to explicitly write out the whole
      type (`OptionSet&lt;Foo&gt;` rather than `Foo`) and the one type using `LOG_IF_DIFFERENT_WITH_CAST_AND_CONSTRUCTION`,
      `SVGPaintOrder`, needs to have a `fromRaw()` static member function.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
    - Update types for `-webkit-line-box-contain` getter/setter/initial functions.
    - Use new `fromRaw` and `toRaw` functions for `SVGPaintOrder`.

* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
    - Remove now unused converters and serializers for `-webkit-line-box-contain`.

* Source/WebCore/style/values/inline/StyleLineBoxContain.h: Removed.
* Source/WebCore/style/values/non-standard/StyleWebKitLineBoxContain.h: Added.
    - Add new strong style type using new `SpaceSeparatedEnumSet` type.

* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+ValueRepresentationNeeded.h:
    - Remove `Style::LineBoxContain`, it no longer exists and its replacement,
      `Style::WebkitLineBoxContainValue`, supports  value representation.

* Source/WebCore/style/values/svg/StyleSVGPaintOrder.h:
    - Add `fromRaw` and `toRaw` functions for use by `RenderStyle`.

Canonical link: <a href="https://commits.webkit.org/302813@main">https://commits.webkit.org/302813@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07f515adaa9b2b6d1187fd832057e6418e3d138e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130284 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41238 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137702 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81854 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132155 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2555 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2447 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99260 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1880 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116672 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79954 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1802 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34800 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80961 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122287 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110345 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140179 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128737 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2345 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2169 "Found 1 new test failure: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/selectors/cue/white-space_normal_wrapped.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107779 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2389 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113015 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107678 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27409 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1843 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31465 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55298 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2415 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65802 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161751 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2232 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40332 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2436 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2341 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
```diff
@@ -116,6 +116,32 @@ template<TupleLike StyleType> requires (std::tuple_size_v<StyleType> == 1) struc
     }
 };
 
+// Specialization for `SpaceSeparatedFixedVector`.
```
Comment is wrong. This is one reason I don't often add comments like this.

darinadler commented on Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
```diff
@@ -116,6 +116,32 @@ template<TupleLike StyleType> requires (std::tuple_size_v<StyleType> == 1) struc
+    }
+};
+
+// Specialization for `CommaSeparatedFixedVector`.
```
Comment is wrong.

weinig committed:
[Style] Convert the '-webkit-line-box-contain' property to strong style types
https://bugs.webkit.org/show_bug.cgi?id=302203

Reviewed by Darin Adler.

Converts the '-webkit-line-box-contain' property to use strong style types.

* Source/WTF/wtf/EnumSet.h:
    - Exposes the underlying enum type via a `using value_type` declaration.
      Needed for use in generic contexts.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new file.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
    - Add mappings macro for `Style::WebkitLineBoxContainValue`.

* Source/WebCore/css/CSSProperties.json:
    - Update style converter used by '-webkit-line-box-contain' to the
      new style type.

* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
    - Adds new aggregates `SpaceSeparatedEnumSet` and `CommaSeparatedEnumSet`
      that wrap `WTF::EnumSet` with specified serialization and conversion.
    - Adds new wrapper, `EnumSetOrKeywordBase` which works like `ValueOrKeyword`
      and `ListOrNone`, except specialized for an `EnumSet` value, meaning it
      has exposed iterators, contains functions, and toRaw/fromRaw functions.

* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBoxBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
* Source/WebCore/rendering/LegacyRootInlineBox.cpp:
* Source/WebCore/rendering/updating/RenderTreeBuilderFirstLetter.cpp:
    - Update for new type name for `-webkit-line-box-contain` values.

* Source/WebCore/rendering/style/RenderStyleDifference.h:
    - Replaces `LOG_IF_DIFFERENT_WITH_CAST_AND_CONSTRUCTION` and `LOG_RAW_OPTIONSET_IF_DIFFERENT`
      with a single `LOG_IF_DIFFERENT_WITH_FROM_RAW`. This means that types using
      `LOG_RAW_OPTIONSET_IF_DIFFERENT` now have to explicitly write out the whole
      type (`OptionSet<Foo>` rather than `Foo`) and the one type using `LOG_IF_DIFFERENT_WITH_CAST_AND_CONSTRUCTION`,
      `SVGPaintOrder`, needs to have a `fromRaw()` static member function.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
    - Update types for `-webkit-line-box-contain` getter/setter/initial functions.
    - Use new `fromRaw` and `toRaw` functions for `SVGPaintOrder`.

* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
    - Remove now unused converters and serializers for `-webkit-line-box-contain`.

* Source/WebCore/style/values/inline/StyleLineBoxContain.h: Removed.
* Source/WebCore/style/values/non-standard/StyleWebKitLineBoxContain.h: Added.
    - Add new strong style type using new `SpaceSeparatedEnumSet` type.

* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+ValueRepresentationNeeded.h:
    - Remove `Style::LineBoxContain`, it no longer exists and its replacement,
      `Style::WebkitLineBoxContainValue`, supports  value representation.

* Source/WebCore/style/values/svg/StyleSVGPaintOrder.h:
    - Add `fromRaw` and `toRaw` functions for use by `RenderStyle`.

Canonical link: https://commits.webkit.org/302813@main
